### ISSUE TITLE
How to insert a new node onto the same line as an existing one

### ISSUE THREAD

**Issue by jacksonrayhamilton** [benjamn/recast#197]
Is it possible to add a node to the following code

```js
'use strict';
```

such that the added node appears on the same line immediately after the previous (when printed)?

```js
'use strict';var a;
```

and similarly such that the node appears before?

```js
var a;'use strict';
```

**Comment by jamestalmage** [comment]
```javascript
var r = require('recast');
var n = r.types.namedTypes;
var b = r.types.builders;
var ast = r.parse('"use strict";');

r.visit(ast,{
  visitExpressionStatement:function(path) {
    var node = path.node;
    if (n.Literal.check(node.expression) &&
      'use strict' === node.expression.value) {
      
      var vDecl = b.variableDeclaration('var',
        [b.variableDeclarator(b.identifier('a'), null)]
      );
      path.insertAfter(vDecl); 
      
      // hack to make printer put it on the same line
      vDecl.loc = {
       start: {
          line: node.loc.end.line,
          column: node.loc.end.column + 1
        },
        end: {
          line: node.loc.end.line,
          column: node.loc.end.column + 1+ 'var a;'.length
        },
        lines: node.loc.lines,
        indent: node.loc.indent
      };
        
    }
    return false;  //don't traverse the tree further
  }
});

console.log(r.print(ast).code);
```

**Comment by jamestalmage** [comment]
swap `insertAfter` for `insertBefore` if you really need it before.

I'd recommend after, since that would make the `loc` values actually correct.

This is totally a hack though, so let @benjamn weigh in before following my advice.

**Comment by benjamn** [comment]
@jamestalmage thanks for taking the time to work out exactly what would be required if you wanted to trick `recast.print` into reprinting unusually formatted new code, rather than reformatting it according to the rules of the pretty printer. I share your feeling that it's hacky, and definitely at odds with Recast's philosophy of taking most of the printing decisions out of the developer's hands. That abstraction is great when you don't care about printing details, but obviously less good when you do care!

@jacksonrayhamilton the printer currently allows a kind of abuse of the AST format, where if it encounters a string instead of an AST node object, it will just print the string as-is. [Here's the code for that](https://github.com/benjamn/recast/blob/f759b1af7c3be5edfdbd3705fc794136292a515f/lib/printer.js#L158-L160).

In other words, I wonder if you could just `recast.print` the `"use strict"` and `var a` nodes yourself, then join the strings together manually, and then insert that combined string as the first "statement" in the block? That might just work. I wouldn't call it elegant, but at least it's straightforward, if it works.

**Comment by zertosh** [comment]
>  the printer currently allows a kind of abuse of the AST format, where if it encounters a string instead of an AST node object,  it will just print the string as-is.

ugh! I wish I had known that sooner. @benjamn is that something you see sticking around or should that be viewed as a hack? With jstransform deprecated, this feature would kinda fill that void.

**Comment by benjamn** [comment]
I definitely plan to continue supporting that <s>hack</s>trick, and maybe even documenting it. My only caution is that it might be risky to insert strings in a transform that runs before other transforms that aren't expecting to find raw strings. But if you control all the transforms, I think this trick is pretty reasonable.

**Comment by benjamn** [comment]
Let me know if the string trick breaks down on you. Pretty sure that's the best answer for now.
